libsolidity/DeclarationContainer.cpp

```diff
@@ -30,6 +30,11 @@ namespace solidity
 
 bool DeclarationContainer::registerDeclaration(Declaration const& _declaration, bool _update)
 {
+	if (_declaration.getName().empty())
+		if (_declaration.getType() != nullptr)
```

Comment:
I think it would be a cleaner solution to not distinguish between functions and other declarations here.
We should rather have a check inside `ContractDefinition::checkTypeRequirements` along these lines:
```
FunctionDefinition const* fallbackFunction = nullptr;
for (ASTPointer<FunctionDefinition> const& function: getDefinedFunctions())
  if (function->getName().empty())
    if (fallbackFunction)
      BOOST_THROW_EXCEPTION(...);
    else
      fallbackFunction = function.get();
```
And then continue with the check that is already there, but do not again test for the scope.